README.md
# 7 Segment Digits Recognization Simple KNN Model

## Sumary

- As request of TTI Viet Nam VSIP2

- KNNs algorithm to train 10-digits model

- Implement computer vision with Python on Raspberry Pi 3 (Debian)

- Threading in Python to handle two cameras in parallel for real-time
  responding

- Build PyQT5 linux application to configure image parameters

## Galery

### Video

[<img title="" src="http://img.youtube.com/vi/vZBaMNVL4jA/0.jpg" alt="demo" data-align="center">](http://www.youtube.com/watch?v=vZBaMNVL4jA "Demo")

### Images

![](resources/img/IMAG0046.jpg)

![](resources/img/IMAG0047.jpg)

![](resources/img/IMAG0048.jpg)

![](resources/img/IMG_20190611_154244.jpg)

## Contributors

1. Nguyễn Minh Tiến Nguyễn Minh Tiến - [Facebook](https://www.facebook.com/spiderock98) | [Telegram](https://t.me/spiderock98) | [Instagram](https://www.instagram.com/spiderock98/) | [YouTube](https://www.youtube.com/channel/UCKtd98ra9ovo2HW4_UFC9Cw/videos)
2. Phạm Chí Tâm - [Facebook](https://www.facebook.com/profile.php?id=100007889464843)
3. Lê Trung Phương - [Facebook](https://www.facebook.com/le.phuong.ltp)




changelog.txt
- dual them upper4digit va lower4digit
- bo if cv2.contourArea(cnt) > 50:

zoom.py
import cv2
import numpy as np
from threading import Thread
import threading


def gamma(image, gamma=1.0):
    if gamma == 0: # The minimal position of the slider is always 0.
        gamma = 1
    table = np.array([((i / 255.0) ** (1.0 / gamma)) * 255
        for i in np.arange(0, 256)]).astype("uint8")
    return cv2.LUT(image, table)


def nothing(x):
    pass


class Zoom:
    def __init__(self, ZoomSize, src, y0, y1, x0, x1):
        self.stream = cv2.VideoCapture(src)
        _, self.frame = self.stream.read()
        self.roi = self.frame[y0:y1, x0:x1]

        cv2.namedWindow('Panel{}'.format(src))
        cv2.createTrackbar('Gamma', 'Panel{}'.format(src), 0, 5, nothing)
        cv2.createTrackbar('Gaussian', 'Panel{}'.format(src), 10, 30, nothing)
        cv2.createTrackbar('Dilate', 'Panel{}'.format(src),1, 10, nothing)
        cv2.createTrackbar('Roi', 'Panel{}'.format(src),36, 100, nothing)

        self.sizeofImage = ZoomSize
        self.src = src
        self.stopped = False
        self.y0, self.y1, self.x0, self.x1 = y0, y1, x0, x1
        self.h, self.w = self.roi.shape[:2]
        self.dim = (self.sizeofImage, int(self.h * (self.sizeofImage / float(self.w))))

    def resize(self, image):
        return cv2.resize(image, self.dim, interpolation=cv2.INTER_AREA)

    def start(self):
        self.e = threading.Event()
        self.e.set() #init True
        t = Thread(target=self.update, args=(self.e,))
        t.daemon = True
        t.start()
        return self

    def update(self, e):
        while e.wait():
            # print('thread {} alive'.format(self.src))
            if self.stopped:
                return

            self.roiBar = cv2.getTrackbarPos('Roi', 'Panel{}'.format(self.src))
            _, self.frame = self.stream.read()
            if self.frame is None:
                continue
            self.roi = self.frame[self.y0:self.y1, self.x0:self.x1]
            self.zoom = self.resize(self.roi)
            self.gammaAdj = gamma(self.zoom, cv2.getTrackbarPos('Gamma', 'Panel{}'.format(self.src)))
            self.barGauss = cv2.getTrackbarPos('Gaussian', 'Panel{}'.format(self.src))
            self.gauss = cv2.GaussianBlur(self.gammaAdj, (self.barGauss * 2 + 1, self.barGauss * 2 + 1), 0)
            self.gray = cv2.cvtColor(self.gauss, cv2.COLOR_BGR2GRAY)
            self.thresh = cv2.adaptiveThreshold(self.gray, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, cv2.THRESH_BINARY_INV, 11, 2)
            self.dst = cv2.dilate(self.thresh, None, iterations=cv2.getTrackbarPos('Dilate', 'Panel{}'.format(self.src)))

    def readDST(self):
        return self.dst

    def readFrame(self):
        return self.frame

    def readZoom(self):
        return self.zoom

    def pause(self):
        self.e.clear()

    def cont(self):
        self.e.set()

    def stop(self):
        self.stopped = True


dataset/train.py
import cv2
import numpy as np
import sys
import imutils

img = cv2.imread('train3.png')
img = imutils.resize(img,900)
gray = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
blur = cv2.GaussianBlur(gray,(7,7),0)
thresh = cv2.adaptiveThreshold(blur,255,1,1,11,2)
# dst = cv2.erode(thresh,None,iterations=1)


#################      Now finding Contours         ###################

contours,_ = cv2.findContours(thresh,cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)

samples =  np.empty((0,100))
responses = []
keys = [i for i in range(48,58)]

# for cnt in contours:
#     if cv2.contourArea(cnt)>50:
#         [x,y,w,h] = cv2.boundingRect(cnt)
#         if  h>10:
#             cv2.rectangle(img,(x,y),(x+w,y+h),(0,0,255),2)

# cv2.drawContours(img, contours, -1, (0,255,0))

# cv2.imshow('img',img)
# cv2.imshow('thresh',thresh)
# cv2.waitKey(0)
# cv2.destroyAllWindows()

for cnt in contours:
    if cv2.contourArea(cnt)>50:
        [x,y,w,h] = cv2.boundingRect(cnt)
        if  h>10:
            cv2.rectangle(img,(x,y),(x+w,y+h),(0,0,255),1)
            roi = thresh[y:y+h,x:x+w]
            roismall = cv2.resize(roi,(10,10))
            cv2.imshow('norm',img)
            key = cv2.waitKey(0)

            if key == 27:  # (escape to quit)
                sys.exit()
            elif key in keys:
                responses.append(int(chr(key)))
                sample = roismall.reshape((1,100))
                samples = np.append(samples,sample,0)

responses = np.array(responses,np.float32)
responses = responses.reshape((responses.size,1))
print ("training complete")

np.savetxt('generalsamples.data',samples)
np.savetxt('generalresponses.data',responses)

0912_digit_dual.py
import cv2
import serial
import numpy as np
from time import sleep
from threading import Thread
import portusb
import cameraindex
import subprocess
from openpyxl import load_workbook, Workbook
import datetime
import os
import shutil
import glob
from zoom import Zoom
from imutils.video import FPS

######################## checking MAC address #########################

# mac=open('/sys/class/net/%s/address' %'eth0').read().strip()
# while( mac!= 'b8:27:eb:df:bd:66'):
#     continue

############################ Training  ############################

samples = np.loadtxt('generalsamples4.data',np.float32)
responses = np.loadtxt('generalresponses4.data',np.float32)
responses = responses.reshape((responses.size,1))
model = cv2.ml.KNearest_create()
model.train(samples, cv2.ml.ROW_SAMPLE, responses)

############################# Init Vars #########################

dctChar3 = dctChar4 = dict() #[vi_tri_x:value]
preRet3 = preRet4 = iFrame3 = iFrame4 = iFail3 = iFail4 = finalReceive = screenRet4 = iInit = lowfinalRec4 = upfinalRec4 = finalRec3 = 0
final3 = final4 = False
flagPass = flagFail = flagSpeed = flagInit = False
seriFrame = 20
# sentFlag = False
# cam4Flag = True

############################# Function  #########################

def newWorkbook():
    global wb,ws
    wb = Workbook()
    ws = wb.active
    # title bar
    timeCol = ws.cell(row=1,column=2,value='Time')
    timeCol.style = 'Title'
    sttCol = ws.cell(row=1,column=3,value='Status')
    sttCol.style = 'Title'
    errRow = ws.cell(row=4,column=6,value='Error')
    errRow.style = 'Calculation'
    totalRow = ws.cell(row=5,column=6,value='Total')
    totalRow.style = 'Calculation'
def insertRow(status):
    global wb,ws
    ws.append([None,datetime.datetime.now().strftime('%d-%m-%Y %H:%M:%S'),status])
    if (status == 'Good'):
        ws['C{}'.format(ws.max_row)].style = 'Good'
    else:
        ws['C{}'.format(ws.max_row)].style = 'Bad'
    wb.save('logs.xlsx')
def filter():
    ws.auto_filter.ref = 'A1:C{}'.format(ws.max_row)
    wb.save('logs.xlsx')

############################# Init I/O #########################

if (os.path.isfile('./logs.xlsx')):
    wb = load_workbook(filename='logs.xlsx')
    ws = wb.active
else:
    newWorkbook()

#port = portusb.port()
port = portusb.portNew()
print('[INFO] Openning ' + port)
ser = serial.Serial(port,9600)

vs3 = Zoom(400,cameraindex.index('Camera'),169,266,189,377).start()
vs4 = Zoom(400,cameraindex.index('CAMERA'),263,372,179,438).start()

sleep(2.0)

fps = FPS().start()

############################# Loop #########################

while True:
    try:
        serInWait = ser.in_waiting
    except OSError:
        vs3.stop()
        vs4.stop()
        ser.close()
        cv2.destroyAllWindows()
        os.system('python3 ./0908_digit_dual.py')
    if (serInWait):
        recieveOrd = [] # reset
        for i in range(13):
            capture = ord(ser.read(1))
            if (capture == 110):
                print('[INFO] Resetting arduino ...')
                screenRet4 = iInit = 0
                final3 = final4 = flagInit = flagFail = flagPass = False
                flagSpeed = True
                ### chi test thu o day ####
                #insertRow('Bad')
            if (capture == 111):
                print('[INFO] Detecting valid USB')
                strUSB = str(subprocess.check_output('lsusb',shell=True))
                #strUSB = glob.glob('/media/pi/[A-Za-z]*')[0]
                if strUSB.rfind('0951:1666') != -1:
                    print('[INFO] USB access grandted')
                    ser.write('USB'.encode())
                    #auto filter
                    ws.auto_filter.ref = 'A1:C{}'.format(ws.max_row)
                    wb.save('logs.xlsx')
                    # cut it out
                    if(os.path.isfile('./logs.xlsx')):
                        shutil.copy('./logs.xlsx', glob.glob('/media/pi/[A-Za-z]*')[0])
                        sleep(5) # waiting for copy
                        os.remove('./logs.xlsx')
                    # then create new wb
                    newWorkbook()
                    ser.write('usb'.encode())
                else:
                    print('[INFO] USB access denied')
            recieveOrd.append(capture)
        # check error nếu bit 3 != 30, 31 và != tổng 3 bit còn lại -> lỗi
        if(recieveOrd[0]+recieveOrd[1]+recieveOrd[2]+recieveOrd[4]+recieveOrd[5]+recieveOrd[6]+recieveOrd[7]+recieveOrd[8]+recieveOrd[9]+recieveOrd[10]+recieveOrd[11])!=recieveOrd[12]:
            ser.write("RESEND\r".encode())
            sleep(0.5)
        finalRec3 = float((recieveOrd[0])*100 + (recieveOrd[1])*10 + (recieveOrd[2]))/100
        lowfinalRec4 = recieveOrd[4]*1000 + recieveOrd[5]*100 + recieveOrd[6]*10 + recieveOrd[7]
        upfinalRec4 = recieveOrd[8]*1000 + recieveOrd[9]*100 + recieveOrd[10]*10 + recieveOrd[11]
        print('[INFO] Thresh: {}'.format(finalRec3))
        print('[INFO] Lower Speed: {} \t Upper Speed: {}'.format(lowfinalRec4, upfinalRec4))

    vs4.cont()
    vs3.pause()
    dst4 = vs4.dst
    roi4 = vs4.zoom
    roiBar4 = vs4.roiBar
    cv2.imshow('Zoom{}'.format(vs4.src), dst4)
    #cv2.imshow('Panel{}'.format(vs4.src), roi4)
    contours4,_= cv2.findContours(dst4,cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)
    dctChar4 = dict() #reset
    for cnt in contours4:
        [x4,y4,w4,h4] = cv2.boundingRect(cnt)
        if (w4 > roiBar4 and h4 > roiBar4):
            _,results,_,_ = model.findNearest(np.float32(cv2.resize(dst4[y4:y4+h4,x4:x4+w4],(10,10)).reshape((1,100))), k = 5)
            dctChar4[x4] = int(results[0][0])
            cv2.rectangle(roi4,(x4,y4),(x4+w4,y4+h4),(0,0,255),2)

    if(len(dctChar4) == 4):
        multi = []
        for xCoor in sorted(dctChar4):
            multi.append(dctChar4[xCoor])
        screenRet4 = multi[0]*1000 + multi[1]*100 + multi[2]*10 + multi[3]
        #if (screenRet4 != preRet4 ):
            #preRet4 = screenRet4
            #print('\t{}'.format(screenRet4))
            
        if ((screenRet4 != preRet4) and (not flagInit)):
            preRet4 = screenRet4
            iInit += 1
            if (iInit > 4):
                iInit = 0
                flagInit = True
                flagSpeed = False
            print('\t{}'.format(screenRet4))
        if (screenRet4 >= lowfinalRec4 and screenRet4 <= upfinalRec4):
            iFrame4 += 1
            #iFail4 = 0
            if (iFrame4 > seriFrame):
                # cam4Flag = False
                # sleep(3)
                # print('[INFO] speed ok')
                # ser.write('SPEED\r'.encode())
                # sleep(0.5)
                final4 = True # red speed ok
                iFrame4 = 0
        else:
            iFrame4=0
        if ((screenRet4 > 1500 and screenRet4 < lowfinalRec4) or screenRet4 > upfinalRec4):
            iFail4 += 1
            #iFrame4 = 0
            if (iFail4 > seriFrame):
                final4 = False
                iFail4 = 0
        else:
            iFail4=0

    vs3.cont()
    vs4.pause()
    dst3 = vs3.dst
    roi3 = vs3.zoom
    roiBar3 = vs3.roiBar
    cv2.imshow('Zoom{}'.format(vs3.src),dst3)
    #cv2.imshow('Panel{}'.format(vs3.src),roi3)
    contours3,_= cv2.findContours(dst3,cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)
    dctChar3 = dict() #reset
    for cnt in contours3:
        [x3,y3,w3,h3] = cv2.boundingRect(cnt)
        if (w3 > roiBar3 and h3 > roiBar3):
            _,results,_,_ = model.findNearest(np.float32(cv2.resize(dst3[y3:y3+h3,x3:x3+w3],(10,10)).reshape((1,100))), k = 5)
            dctChar3[x3] = int(results[0][0])
            cv2.rectangle(roi3,(x3,y3),(x3+w3,y3+h3),(0,0,255),2)

    if (len(dctChar3) == 3):
        multi = []
        for xCoor in sorted(dctChar3):
            multi.append(dctChar3[xCoor])
        screenRet3 = float(multi[0]*100 + multi[1]*10 + multi[2])/100
        if (screenRet3 != preRet3):
            preRet3 = screenRet3
            print(screenRet3)
        if (screenRet3 <= finalRec3):
            iFail3 = 0
            iFrame3 += 1
            if (iFrame3 > seriFrame):
                # print('[INFO] PASS')
                # insertRow('Good')
                # ser.write("PASS\r".encode()) # green light
                # sleep(0.5)
                final3 = True
                iFrame3 = 0
        else:
            iFrame3 = 0
            iFail3 += 1
            if (iFail3 > seriFrame):
                final3 = False
                iFail3 = 0

    if (final3 and final4 and not flagPass):
        flagPass = True
        flagSpeed = False
        print('[INFO] Pass')
        ser.write('PASS\r'.encode())
    elif (screenRet4 <= 1500 and not flagFail):
        flagFail = flagSpeed = True
        flagPass = flagInit = False
        print('[INFO] WHITE')
        ser.write('WHITE\r'.encode())
    elif (not final3 and final4 and not flagSpeed):
        flagSpeed = True
        flagPass = False
        print('[INFO] Failed 1')
        ser.write('SPEED\r'.encode())
    elif (final3 and not final4 and not flagSpeed):
        flagSpeed = True
        flagPass = False
        print('[INFO] Failed 2')
        ser.write('SPEED\r'.encode())
    elif (not final3 and not final4 and not flagSpeed):
        flagSpeed = True
        flagPass = False
        print('[INFO] Failed 3')
        ser.write('SPEED\r'.encode())
    
    if cv2.waitKey(1) == 27:
        break
    fps.update()

fps.stop()
print('[INFO] FPS: {:.2f}'.format(fps.fps()))
vs3.stop()
vs4.stop()
ser.close()
cv2.destroyAllWindows()


fast.py
import cv2
import numpy as np 
# import sys
import imutils
from imutils.video import VideoStream
from time import sleep
from imutils.video import FPS
import cameraindex

######################## checking MAC address #########################

# mac=open('/sys/class/net/%s/address' %'eth0').read().strip()
# while( mac!= 'b8:27:eb:df:bd:66'):
#     continue

############################ Training  ############################

samples = np.loadtxt('generalsamples4.data',np.float32)
responses = np.loadtxt('generalresponses4.data',np.float32)
responses = responses.reshape((responses.size,1))
model = cv2.ml.KNearest_create()
model.train(samples, cv2.ml.ROW_SAMPLE, responses)

############################# Init Vars #########################

refPt = [(0,0),(0,0)]
cropping = False

############################# Function  #########################
def nothing(x):
    pass

def gamma(image, gamma=1.0):
    if gamma == 0: # The minimal position of the slider is always 0.
        gamma = 1
    invGamma = 1.0 / gamma
    table = np.array([((i / 255.0) ** invGamma) * 255
        for i in np.arange(0, 256)]).astype("uint8")
    return cv2.LUT(image, table)

def click_and_crop(event, x, y, flags, param):
    global refPt, cropping
    if event == cv2.EVENT_LBUTTONDOWN:
        refPt = [(x,y),(0,0)]
        cropping = True
        
    elif event == cv2.EVENT_LBUTTONUP:
        refPt[1] = (x,y)
        cropping = False
        #print('xUp={}  yUp={}'.format(refPt[0][0],refPt[0][1]))
        #print('xDown={}  yDown={}\n'.format(refPt[1][0],refPt[1][1]))
        print('{},{},{},{}'.format(refPt[0][1],refPt[1][1],refPt[0][0],refPt[1][0]))

############################# Init I/O #########################

cv2.namedWindow("image")
cv2.namedWindow("Panel")

cv2.createTrackbar('Gamma','Panel',5,5,nothing)
cv2.createTrackbar('Gaussian','Panel',13,30,nothing)
cv2.createTrackbar('Dilate','Panel',0,10,nothing)
cv2.createTrackbar('Roi','Panel',20,100,nothing)

cv2.setMouseCallback('image', click_and_crop)

vs = VideoStream(cameraindex.index('CAMERA')).start()
vs = VideoStream(cameraindex.index('Camera')).start()

sleep(1.0)

fps = FPS().start()

imgSize = 500
############################# Loop  #########################

while True:
    im = vs.read()
    ori = im.copy()
    gammaAdj = gamma(ori,cv2.getTrackbarPos('Gamma','Panel'))
    im = gammaAdj[refPt[0][1]+1:refPt[1][1]-1,refPt[0][0]+1:refPt[1][0]-1]
    zoom = imutils.resize(im,imgSize)

    gray = cv2.cvtColor(zoom,cv2.COLOR_BGR2GRAY)
    barGauss = cv2.getTrackbarPos('Gaussian','Panel')
    gauss = cv2.GaussianBlur(gray, (barGauss*2+1,barGauss*2+1), 0)
    thresh = cv2.adaptiveThreshold(gauss,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY_INV,11,2)
    barDilEro = cv2.getTrackbarPos('Dilate','Panel')
    dst = cv2.dilate(thresh,None,iterations=barDilEro*2+1)
    contours,_= cv2.findContours(dst,cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)

    for cnt in contours:
        if cv2.contourArea(cnt)>50:
            [x,y,w,h] = cv2.boundingRect(cnt)
            wh = cv2.getTrackbarPos('Roi','Panel')
            #if  ((w>wh-20) and (h>wh)):
            if  (h>wh):
                cv2.rectangle(zoom,(x,y),(x+w,y+h),(0,255,0),2)
                roi = dst[y:y+h,x:x+w]
                roismall = cv2.resize(roi,(10,10))
                roismall = roismall.reshape((1,100))
                roismall = np.float32(roismall)
                retval, results, neigh_resp, dists = model.findNearest(roismall, k = 5)
                val = (int((results[0][0])))
                cv2.putText(zoom,str(val),(x,y+w),0,fontScale=1,color=(0,0,255), thickness=2)

    fps.update()

    cv2.imshow('Panel',dst)
    cv2.imshow('image',ori)
    cv2.imshow('zoom',zoom)

    if cv2.waitKey(1) == 27:
        break

fps.stop()
print('FPS: {:.2f}'.format(fps.fps()))

vs.stop()
cv2.destroyAllWindows()
